### PR TITLE
Spark: value list of IN/NOT_IN containing null value should not be converted to Iceberg expression

### PR DESCRIPTION
**Pull request by ConeyLiu**
Address the comments: https://github.com/apache/iceberg/pull/7886#discussion_r1280860230

### CODE REVIEW

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java [hunk 1]
```diff
@@ -161,10 +161,13 @@ public static Expression convert(Filter filter) {
 
         case IN:
           In inFilter = (In) filter;
+          Preconditions.checkArgument(
+              Stream.of(inFilter.values()).noneMatch(Objects::isNull),
+              "Expression can not be converted (in is not null-safe): %s",
+              filter);
```
Comment (rdblue) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java:167]: This should return null, not throw an exception. We don't ever want to fail conversion, just return `null` so the caller doesn't push an expression down.

**Review by rdblue — CHANGES_REQUESTED** [review]
@ConeyLiu, I think this takes the wrong approach. If an expression cannot be converted, the conversion should return `null` rather than throwing an exception.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java (hunk 1 above)
Comment (ConeyLiu) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java]: Updated

**Comment by ConeyLiu** [comment]
Thanks @rdblue for the reviewing. I want to keep the same behavior as the null processing for `eq`. Changed to return null because it is more reasonable.

**Review by rdblue — APPROVED** [review]
This looks correct to me, but I would appreciate it if @aokolnychyi could take a look at it as well.

**Comment by aokolnychyi** [comment]
Let me check in a bit.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java [hunk 2]
```diff
@@ -161,10 +162,13 @@ public static Expression convert(Filter filter) {
 
         case IN:
           In inFilter = (In) filter;
+          if (Stream.of(inFilter.values()).anyMatch(Objects::isNull)) {
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java:165]: Why is the previous logic wrong? 

```
col1 IN (1, 2, NULL) -> gives true if col1 is 1
col2 IN (1, 2, NULL) -> gives null if col1 is 10 and null means no match
col3 IN (1, 2, NULL) -> gives null if col1 is null and null means no match
```

Nulls in predicates are handled in a special way as they are same as false. 
There is `ReplaceNullWithFalseInPredicate` in Spark that I worked on 4-5 years ago.

Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java:165]: We previously removed nulls as Iceberg filters behave as Java filters while Spark passes SQL filters.

Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java:165]: I am sorry missed the initial discussion. @rdblue @ConeyLiu, could you elaborate?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java [hunk 3]
```diff
@@ -178,6 +182,10 @@ public static Expression convert(Filter filter) {
             // col NOT IN (1, 2) in Spark is equivalent to notNull(col) && notIn(col, 1, 2) in
             // Iceberg
             In childInFilter = (In) childFilter;
+            if (Stream.of(childInFilter.values()).anyMatch(Objects::isNull)) {
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java:185]: Is it even realistic to get such filters past the optimizer? Won't they be replaced with false?

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/sql/TestFilterPushDown.java [hunk 4]
```diff
@@ -99,8 +99,14 @@ public void testFilterPushdownWithIdentityTransform() {
...
     checkOnlyIcebergFilters(
+        "dep IN ('d1', 'd2')" /* query predicate */,
+        "dep IN ('d1', 'd2')" /* Iceberg scan filters */,
+        ImmutableList.of(row(1, 100, "d1"), row(2, 200, "d2")));
+
+    checkFilters(
         "dep IN (null, 'd1')" /* query predicate */,
-        "dep IN ('d1')" /* Iceberg scan filters */,
+        "dep IN (null,d1)" /* Spark post scan filter */,
+        "" /* Iceberg scan filters */,
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/sql/TestFilterPushDown.java:109]: This does not seem right. Why can't we push a predicate like this?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java (hunk 3 above)
Comment (rdblue) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java:185]: That's one reason why we don't want to bother. If it has a NULL something is weird.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java (hunk 2 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java:165]: I thought we handled such cases in a special way. For instance, there is `hasNoInFilter` used in the negation to recursively check for nested NOT IN inside NOT. We do handle IN and NOT IN differently.

My worry is that filters like `IN (1, 2, NULL)`, which are perfectly fine to push down, will no longer be pushed down, causing silent performance issues when predicates are computes programatically.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java (hunk 3 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java:185]: If we want to be safer and not support such NOT IN predicates, I can see that. I am not yet convinced about IN predicates, however.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java (hunk 2 above)
Comment (rdblue) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java:165]: Ah, I missed that we have specific handling to ensure that there is no IN nested within NOT. I think you're right that this is okay after all. We should probably double check more cases, like that the handling for `NOT(IN(1, 2))` is also correct for `NOT(IN(1, 2, null)`.

**Comment by rdblue** [comment]
I'm going to close this because @aokolnychyi is right: we handle IN this way because we have special checks within the NOT conversion that ensure we don't negate an IN expression. Thanks, Anton!

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java (hunk 2 above)
Comment (ConeyLiu) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java:165]: Thanks @aokolnychyi @rdblue for the explanations. So the reason that we can push down safety is:
1. `NULL` in a predicate will be optimized into false in Spark.
2. We have special processing for `NOT IN` and don't allow other `IN` nested within `NOT`. The Spark `col NOT IN (1, 2)` is converted to Iceberg expressions: `notNull(col) && notIn(col, 1, 2)`.

I have a question about `IN`. Is it safe to convert `col IN (1, 2)` to Iceberg: `notNull(col) && in(col, 1, 2)` as well? Why not do that?

**Pull request closed without merging**